src/libs/elektra/keyset.c

```diff
@@ -1254,6 +1254,83 @@ elektraCursor ksFindHierarchy (const KeySet * ks, const Key * root, elektraCurso
 	return it;
 }
 
+/**
+ * Retrieves all Keys from KeySet @p ks that are below or at @p root
+ *
+ * This function is similar to ksCut().
+ * It returns an identical KeySet, but the Keys also remain in @p ks
+ *
+ * @param ks the Keyset to copy from
+ * @param root the point where to copy from the Keyset
+ *
+ * @return a new allocated KeySet which needs to deleted with ksDel().
+ *         The KeySet consists of all Keys (of the original KeySet ks)
+ *         below @p root. If @p root exists, it will also be appended.
+ */
+KeySet * ksBelow (const KeySet * ks, const Key * root)
```

Comment:
Don't really like this function if there is already ksCut. Why not simply cut and reappend? Reappending could be done quite efficient with sorting.

Reply:
Well this function allows us to work on `const` KeySets.

Reply:
1. As stated already the keyset could read-only.
2. Cutting and reappending is a stupid wasteful operation that might be somewhat efficient because of the sorting, but still involves multiple `memmove` of potentially large arrays (depending on the keyset).
3. I also don't like having both `ksCut` and `ksBelow` (and also `ksFindHierarchy`). If I could decide I'd have:
   ```c
   // exactly as it exists
   elektraCursor ksFindHierarchy (const KeySet * ks, const Key * root, elektraCursor * end);

   // copies all keys between two positions into a new keyset
   KeySet * ksCopyRange (const KeySet * ks, elektraCursor start, elektraCursor end);

   // delete all keys between two positions
   void ksRemoveRange (KeySet * ks, elektraCursor start, elektraCursor end);
   ```
   This covers the existing cases and more (because copying/removing is not tied to hierarchy). The replacements would be:
   ```c
   static inline KeySet * ksBelow (const KeySet * ks, const Key * root) {
       elektraCursor end;
       elektraCursor start = ksFindHierarchy (ks, root, &end);
       return ksCopyRange (ks, start, end);
   }
   
   static inline KeySet * ksCut (KeySet * ks, const Key * cutpoint) {
       elektraCursor end;
       elektraCursor start = ksFindHierarchy (ks, root, &end);
       KeySet * cut = ksCopyRange (ks, start, end);
       ksRemoveRange (ks, start, end);
       return cut;
   }
   ```

Reply:
Yeah, we won't find a solution now so let us temporarily keep it as-is.